src/profile-logic/sanitize.js

```diff
@@ -395,9 +446,16 @@ function sanitizeThreadPII(
     const { frameTable, funcTable, resourceTable, stackTable, samples } =
       newThread;
     for (let frameIndex = 0; frameIndex < frameTable.length; frameIndex++) {
-      const innerWindowId = frameTable.innerWindowID[frameIndex];
+      const innerWindowID = frameTable.innerWindowID[frameIndex];
       const funcIndex = frameTable.func[frameIndex];
-      if (innerWindowId !== null && windowIdToBeSanitized.has(innerWindowId)) {
+
+      const isPrivateBrowsing =
+        innerWindowID && windowIdFromPrivateBrowsing.has(innerWindowID);
+      const isRemoveTabId =
+        innerWindowID &&
+        PIIToBeRemoved.shouldKeepTabID !== null &&
+        !windowIdFromActiveTab.has(innerWindowID);
```

Comment:
It's not exactly that.
There are 2 distinct (but related) things happening when sanitizing the sample data:
1. sanitizing the frames (== the function names): we sanitize them only if they have innerWindowID information that's different than the ones in the active tab id. This way we keep the function names in the stacks. I think this doesn't go as far as I'd want, because I believe we should sanitize all frames that are only found in stacks where all frames either have no innerWindowID or have one that's not part of the tab.
2. removing the sample data: for these we remove them also when they don't have an innerWindowID. This is consistent with the view we show in the active tab view.

Let's take a few examples.

1. Stack with JS frames: [main, a.js:callDom, addEventListener, addEventListenerCppImplementation]
   a. in the tab
   b. in another tab
2. Stack without JS frames: [main, RefreshDriverTick, Paint https://cnn.com]. This is a simplified version of a stack I found, where a Paint label frame contains the name of the website. This has no innerWindowID anywhere. (but maybe Paint label frames should have them, but that's another story).

1a: we don't want to sanitize any of these frames, nor remove any sample with this stack. The innerWindowID information from the JS frame is inherited to the end of the stack.
1b: we want to sanitize these frames only if they're not used in another stack. But we don't do that at the moment, we only sanitize the JS frames and don't inspect stacks. Therefore some PII values will still be in the stringTable.
But we'll remove all samples with this stack.
2: we don't sanitize these frames (but we probably should, at least for the Paint label frame). The Paint label frame could be further sanitized with the "remove URL" checkbox, but the user doesn't know about that really (but it's enabled by default for release). We remove all samples for this stack because it contains no innerWindowID information anywhere.

At the end, we're checking if any stack or any marker are left, and because we're removing all these samples, in a fission Firefox, we usually then remove the full thread, which of course removes also the stringTable in that case, as well as the thread name which contains the website URL. If we don't remove the samples as agressively, most probably a few samples would still be here (I've seen it in real profiles) and we wouldn't be able to remove the thread, and the user wouldn't even know because it's not visible in the active tab view.

I hope that with these explanations the behavior is clearer as well as its rationale.

Reply:
Yeah, it's clearer now. This sounds reasonable to me!